sql/src/main/java/io/crate/planner/fetch/MultiSourceFetchPushDown.java

```diff
@@ -57,46 +65,54 @@
 
     void process() {
         remainingOutputs = statement.querySpec().outputs();
-        statement.querySpec().outputs(new ArrayList<>());
 
         HashMap<Symbol, FetchReference> fetchRefByOriginalSymbol = new HashMap<>();
         ArrayList<Symbol> mssOutputs = new ArrayList<>(
             statement.sources().size() + statement.requiredForQuery().size());
 
-        for (Map.Entry<QualifiedName, RelationSource> entry : statement.sources().entrySet()) {
-            RelationSource source = entry.getValue();
-            if (!(source.relation() instanceof DocTableRelation)) {
-                mssOutputs.addAll(source.querySpec().outputs());
+        for (Map.Entry<QualifiedName, AnalyzedRelation> entry : statement.sources().entrySet()) {
+            QueriedRelation relation = (QueriedRelation) entry.getValue();
+            if (!(relation instanceof QueriedDocTable)) {
+                mssOutputs.addAll(relation.fields());
                 continue;
             }
 
-            DocTableRelation rel = (DocTableRelation) source.relation();
-            HashSet<Field> canBeFetched = filterByRelation(statement.canBeFetched(), rel);
+            DocTableRelation rel = ((QueriedDocTable) relation).tableRelation();
+            DocTableInfo tableInfo = rel.tableInfo();
+            FetchFields canBeFetched = filterByRelation(statement.canBeFetched(), relation, rel);
             if (!canBeFetched.isEmpty()) {
 
                 Field fetchIdColumn = rel.getField(DocSysColumns.FETCHID);
+                assert fetchIdColumn != null: "_fetchId must be accessible";
                 mssOutputs.add(fetchIdColumn);
-                InputColumn fetchIdInput = new InputColumn(mssOutputs.size() - 1);
+                InputColumn fetchIdInput = new InputColumn(mssOutputs.size() - 1, fetchIdColumn.valueType());
 
                 ArrayList<Symbol> qtOutputs = new ArrayList<>(
-                    source.querySpec().outputs().size() - canBeFetched.size() + 1);
+                    relation.querySpec().outputs().size() - canBeFetched.size() + 1);
                 qtOutputs.add(fetchIdColumn);
 
-                for (Symbol output : source.querySpec().outputs()) {
+                for (int i = 0; i < relation.querySpec().outputs().size(); i++) {
+                    Symbol output = relation.querySpec().outputs().get(i);
                     if (!canBeFetched.contains(output)) {
                         qtOutputs.add(output);
-                        mssOutputs.add(output);
+                        mssOutputs.add(relation.fields().get(i));
                     }
                 }
-                for (Field field : canBeFetched) {
+                for (Tuple<Field, Reference> parentAndChild : canBeFetched.parentAndChildren()) {
```

Comment:
Please some comment to explain what's happening here (and block after the for loop). thx!